e2e/testcases/reconciler_finalizer_test.go

```diff
@@ -492,6 +498,88 @@ func TestReconcilerFinalizer_MultiLevelMixed(t *testing.T) {
+	namespaceNN := types.NamespacedName{Name: "managed-ns"}
+	contrivedFinalizer := "e2e-test"
+	nt := nomostest.New(t, nomostesting.MultiRepos,
+		ntopts.Unstructured,
+		ntopts.RootRepo(nestedRootSyncNN.Name),      // Create a nested RootSync to delete mid-test
+		ntopts.WithCentralizedControl,               // This test assumes centralized control
+		ntopts.WithReconcileTimeout(10*time.Second), // Reconcile expected to fail, so use a short timeout
+	)
+	// add a Namespace to the nested RootSync
+	namespace := fake.NamespaceObject(namespaceNN.Name)
+	nsPath := nomostest.StructuredNSPath(namespace.GetNamespace(), namespaceNN.Name)
+	nt.Must(nt.RootRepos[nestedRootSyncNN.Name].Add(nsPath, namespace))
+	nt.Must(nt.RootRepos[nestedRootSyncNN.Name].CommitAndPush(fmt.Sprintf("add Namespace %s", namespaceNN.Name)))
+	if err := nt.WatchForAllSyncs(); err != nil {
+		nt.T.Fatal(err)
+	}
+	// Add a fake finalizer to the namespace to block deletion
+	testutils.WithFinalizer(namespace, "config-sync/e2e-test")
+	nt.T.Logf("Add a fake finalizer named %s to Namespace %s", contrivedFinalizer, namespaceNN.Name)
+	if err := nt.KubeClient.Apply(namespace); err != nil {
+		nt.T.Fatal(err)
+	}
+	// Try to remove the nested RootSync. Deletion should be blocked by ns finalizer
+	nestedRootSyncPath := fmt.Sprintf("acme/namespaces/%s/%s.yaml",
+		configsync.ControllerNamespace, nestedRootSyncNN.Name)
+	nt.Must(nt.RootRepos[rootSyncNN.Name].Remove(nestedRootSyncPath))
+	nt.Must(nt.RootRepos[rootSyncNN.Name].CommitAndPush(fmt.Sprintf("remove Namespace %s", namespaceNN.Name)))
+	expectedCondition := &v1beta1.RootSyncCondition{
+		Type:    v1beta1.RootSyncReconcilerFinalizerFailure,
+		Status:  "True",
+		Reason:  "DestroyFailure",
+		Message: "Failed to delete managed resource objects",
+		Errors: []v1beta1.ConfigSyncError{
+			{
+				Code:         applier.ApplierErrorCode,
+				ErrorMessage: "KNV2009: failed to wait for Namespace, /managed-ns: reconcile timeout\n\nFor more information, see https://g.co/cloud/acm-errors#knv2009",
+			},
+		},
+	}
+	// Finalizer currently only sets condition, not sync status
+	if err := nt.Watcher.WatchObject(kinds.RootSyncV1Beta1(), nestedRootSyncNN.Name, nestedRootSyncNN.Namespace,
+		[]testpredicates.Predicate{testpredicates.RootSyncHasCondition(expectedCondition)}); err != nil {
+		nt.T.Fatal(err)
+	}
+	// Wait a fixed duration for RootSync deletion to be blocked
+	time.Sleep(30 * time.Second)
+	if err := nt.Validate(nestedRootSyncNN.Name, nestedRootSyncNN.Namespace, &v1beta1.RootSync{}); err != nil {
+		nt.T.Fatal(err)
+	}
+	if err := nt.Validate(namespaceNN.Name, namespaceNN.Namespace, &corev1.Namespace{}); err != nil {
+		nt.T.Fatal(err)
+	}
+	// Remove the fake finalizer
+	namespace = fake.NamespaceObject(namespaceNN.Name)
+	testutils.WithoutFinalizers(namespace)
+	nt.T.Logf("Remove the fake finalizer named %s from Namespace %s", contrivedFinalizer, namespaceNN.Name)
+	if err := nt.KubeClient.Apply(namespace); err != nil {
```

Comment:
SSA probably works here, but only because you're setting the finalizers to an empty list, instead of nil. If you set it to nil, you'd have to be using an Unstructured object to get it to pass the nil instead of skipping the field (due to omitempty on the Namespace struct field).

It might be clearer just to use a merge Patch with JSON. There's lots of examples in other tests.

Reply:
Actually... with SSA, since you already appended a finalizer, it should remove it simply by not passing it. So it should be safe to just replace the empty list with a nil.